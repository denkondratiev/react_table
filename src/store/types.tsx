import {
  setParams,
  setTable,
  setRows,
  setCells,
  setShowButtons,
  increment,
  setNewRow,
  removeRow
} from './actions'

export const ACTION_TYPES = {
  SET_PARAMS: 'SET::PARAMS',
  SET_TABLE: 'SET::TABLE',
  SET_ROWS: 'SET::ROWS',
  SET_CELLS: 'SET::CELLS',
  SHOW_BUTTONS: 'SHOW::BUTTONS',
  INCREMENT: 'INCREMENT',
  ADD_ROW: 'ADD::ROW',
  REMOVE_ROW: 'REMOVE::ROW'
}

export type State = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
  params: Params
  buttons: boolean
}

export type Action = (
  | ReturnType<typeof setParams>
  | ReturnType<typeof setTable>
  | ReturnType<typeof setRows>
  | ReturnType<typeof setCells>
  | ReturnType<typeof setShowButtons>
  | ReturnType<typeof increment>
  | ReturnType<typeof setNewRow>
  | ReturnType<typeof removeRow>
)

export type Params = {
  [name: string]: string | number
}

export type RowsParams = {
  [name: string]: Array<string>
}

export type CellsParams = {
  [name: string]: { id: string, amount: number }
}

export type NewRowsParams = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
}
