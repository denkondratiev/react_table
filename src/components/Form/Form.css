form {
  position: relative;
  display: block;
  max-width: 200px;
  margin: 0 auto;

  padding: 30px 0;
}

.form-control {
  margin-right: 20px;
  margin-bottom: 20px;
}

.form-control:last-child {
  margin-right: 0;
}

.error {
  border: 1px solid red !important;
}
