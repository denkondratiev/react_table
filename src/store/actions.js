export const ACTION_TYPES = {
  SET_COLUMNS: 'SET::COLUMNS',
  SET_ROWS: 'SET::ROWS',
  SET_TABLE: 'SET::TABLE',
  SET_HIGHLIGHT: 'SET::HIGHLIGHT',
  SET_SHOW_TABLE: 'SET::SHOW::TABLE',
  SET_SHOW_BUTTONS: 'SET::SHOW::BUTTONS'
}

export const setColumns = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_COLUMNS
})

export const setRows = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_ROWS
})

export const setHighlights = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_HIGHLIGHT
})

export const setTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_TABLE
})

export const setShowTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_TABLE
})

export const setShowButtons = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_BUTTONS
})
