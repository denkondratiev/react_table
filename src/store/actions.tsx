import {
  ACTION_TYPES,
  Params,
  RowsParams,
  CellsParams,
  NewRowsParams
} from './types'

export const setParams = (params: Params) => ({
  payload: params,
  type: ACTION_TYPES.SET_PARAMS
})

export const setTable = (params: Array<string>) => ({
  payload: params,
  type: ACTION_TYPES.SET_TABLE
})

export const setRows = (params: RowsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_ROWS
})

export const setCells = (params: CellsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_CELLS
})

export const setShowButtons = (params: boolean) => ({
  payload: params,
  type: ACTION_TYPES.SHOW_BUTTONS
})

export const increment = (params: string) => ({
  payload: params,
  type: ACTION_TYPES.INCREMENT
})

export const setNewRow = (params: NewRowsParams) => ({
  payload: params,
  type: ACTION_TYPES.ADD_ROW
})

export const removeRow = (params: any) => ({
  payload: params,
  type: ACTION_TYPES.REMOVE_ROW
})
