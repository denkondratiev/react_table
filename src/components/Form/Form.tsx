import React, { useState } from 'react'
import { generateTable } from '../../helpers/generate'
import { useDispatch, batch } from 'react-redux'
import './Form.css'
import {
  setParams,
  setTable,
  setRows,
  setCells,
  setShowButtons
} from '../../store/actions'

type InputValue = {
  rowsAmount: string,
  columnsAmount: string,
  lightsAmount: string
}

type ErrorValue = {
  rowsAmount: boolean,
  columnsAmount: boolean,
  lightsAmount: boolean
}

const Form: React.FC = () => {
  const dispatch = useDispatch()

  const [inputValue, setInputValue] = useState<InputValue>({
    rowsAmount: '',
    columnsAmount: '',
    lightsAmount: ''
  })

  const [error, setError] = useState<ErrorValue>({
    rowsAmount: false,
    columnsAmount: false,
    lightsAmount: false
  })

  const handleGenerateTable = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault()

    const rowsAmount = inputValue.rowsAmount
    const columnsAmount = inputValue.columnsAmount
    const lightsAmount = inputValue.lightsAmount

    if (Number(rowsAmount) > 0 && Number(columnsAmount) > 0 && Number(lightsAmount) > 0) {
      const { table, rows, cells } = generateTable(rowsAmount, columnsAmount)

      batch(() => {
        dispatch(setParams({ rowsAmount, columnsAmount, lightsAmount }))
        dispatch(setTable(table))
        dispatch(setRows(rows))
        dispatch(setCells(cells))
        dispatch(setShowButtons(true))
      })

      setInputValue({
        rowsAmount: '',
        columnsAmount: '',
        lightsAmount: ''
      })
    }

    setError({
      rowsAmount: !!((!rowsAmount || Number(rowsAmount) <= 0)),
      columnsAmount: !!((!columnsAmount || Number(columnsAmount) <= 0)),
      lightsAmount: !!((!lightsAmount || Number(lightsAmount) <= 0))
    })
  }

  const onChangeHandler = (event: React.FormEvent<HTMLInputElement>): void => {
    const { name, value } = event.currentTarget

    setError({ ...error, [name]: false })
    setInputValue({ ...inputValue, [name]: value })
  }

  return (
    <form
      id="form"
      className="form"
      onSubmit={(event) => handleGenerateTable(event)}>
      {
        error.rowsAmount && (
          <div>
              Please add correct rows value
          </div>
        )
      }
      <input
        type="number"
        name="rowsAmount"
        className={`form-control ${error.rowsAmount && 'error'}`}
        placeholder="Rows..."
        onChange={(event) => onChangeHandler(event)}
        value={inputValue.rowsAmount}
      />
      {
        error.columnsAmount && (
          <div>
              Please add correct columns value
          </div>
        )
      }
      <input
        type="number"
        name="columnsAmount"
        className={`form-control ${error.columnsAmount && 'error'}`}
        placeholder="Columns..."
        onChange={(event) => onChangeHandler(event)}
        value={inputValue.columnsAmount}
      />
      {
        error.lightsAmount && (
          <div>
              Please add correct highlight value
          </div>
        )
      }
      <input
        type="number"
        name="lightsAmount"
        className={`form-control ${error.lightsAmount && 'error'}`}
        placeholder="Highlight cells..."
        onChange={(event) => onChangeHandler(event)}
        value={inputValue.lightsAmount}
      />
      <button
        id ="button-generate"
        type="submit"
        className="btn btn-dark"
      >
        Generate
      </button>
    </form>
  )
}

export default Form
