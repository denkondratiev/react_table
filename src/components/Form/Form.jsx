import React, { useState } from 'react'
import { FormShape } from '../../helpers/shapes'
import { generateMatrix } from '../../helpers/helpers'
import { connect } from 'react-redux'
import store from '../../store/reducers'
import './Form.css'
import {
  setColumns,
  setRows,
  setHighlights,
  setTable,
  setShowTable,
  setShowButtons
} from '../../store/actions'

const Form = (props) => {
  const { rows, columns, highlights } = props

  const [isGenerate, setIsGenerate] = useState(false)
  const [error, setError] = useState({
    rows: false,
    columns: false,
    highlights: false
  })

  const handleGenerateTable = (event) => {
    event.preventDefault()

    if (rows && columns && highlights) {
      const matrix = generateMatrix(rows, columns)

      store.dispatch(setTable(matrix))
      store.dispatch(setShowTable(true))
      store.dispatch(setShowButtons(true))

      setIsGenerate(true)
    }

    const errorObj = {
      rows: null,
      columns: null,
      highlight: null
    }

    if (!rows) {
      errorObj.rows = true
    }

    if (!columns) {
      errorObj.columns = true
    }

    if (!highlights) {
      errorObj.highlights = true
    }

    setError({
      rows: errorObj.rows,
      columns: errorObj.columns,
      highlights: errorObj.highlights
    })
  }

  const setColumnsHandler = ({ name, value }) => {
    if (value === '') {
      store.dispatch(setColumns(''))

      return false
    }

    setError({ columns: false })
    store.dispatch(setColumns(value.replace(/[^0-9]/, '')))
  }

  const setRowsHandler = ({ name, value }) => {
    if (value === '') {
      store.dispatch(setRows(''))

      return false
    }

    setError({ rows: false })
    store.dispatch(setRows(value.replace(/[^0-9]/, '')))
  }

  const setHighlightsHandler = ({ name, value }) => {
    if (value === '') {
      store.dispatch(setHighlights(''))

      return false
    }

    setError({ highlights: false })
    store.dispatch(setHighlights(value.replace(/[^0-9]/, '')))
  }

  return (
    <form className="form" onSubmit={(event) => handleGenerateTable(event)}>
      {
        error.rows && (
          <div>
              Please add correct rows value
          </div>
        )
      }
      <input
        type="number"
        name="rows"
        className={`form-control ${error.rows && 'error'}`}
        placeholder="Rows..."
        onChange={(event) => setRowsHandler(event.target)}
        value={isGenerate ? '' : rows}
      />
      {
        error.columns && (
          <div>
              Please add correct columns value
          </div>
        )
      }
      <input
        type="number"
        name="columns"
        className={`form-control ${error.columns && 'error'}`}
        placeholder="Columns..."
        onChange={(event) => setColumnsHandler(event.target)}
        value={isGenerate ? '' : columns}
      />
      {
        error.highlights && (
          <div>
              Please add correct highlight value
          </div>
        )
      }
      <input
        type="number"
        name="highlights"
        className={`form-control ${error.highlights && 'error'}`}
        placeholder="Highlight cells..."
        onChange={(event) => setHighlightsHandler(event.target)}
        value={isGenerate ? '' : highlights}
      />
      <button
        type="submit"
        className="btn btn-dark"
      >
        Generate
      </button>
    </form>
  )
}

Form.propTypes = FormShape.isRequired

const mapStateToProps = state => ({
  columns: state.columns,
  rows: state.rows,
  highlights: state.highlights
})

export default connect(mapStateToProps)(Form)
