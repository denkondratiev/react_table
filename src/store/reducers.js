import { createStore } from 'redux'
import { ACTION_TYPES } from './actions'

function rootReducer (state, action) {
  const {
    SET_COLUMNS,
    SET_ROWS,
    SET_HIGHLIGHTS,
    SET_TABLE,
    SET_SHOW_TABLE,
    SET_SHOW_BUTTONS,
    INCREMENT,
    SET_LIGHT_CELLS
  } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE: {
      return {
        ...state,
        table: action.payload
      }
    }

    case SET_COLUMNS: {
      return {
        ...state,
        columns: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_ROWS: {
      return {
        ...state,
        rows: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_HIGHLIGHTS: {
      return {
        ...state,
        highlights: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_SHOW_TABLE: {
      return {
        ...state,
        showTable: action.payload
      }
    }

    case SET_SHOW_BUTTONS: {
      return {
        ...state,
        showButtons: action.payload
      }
    }

    case INCREMENT: {
      state.table.splice(action.payload.id, 1, [...state.table[action.payload.id]
        .map((item, index) => index !== action.payload.cellIndex
          ? item
          : {
            ...state.table[action.payload.id][action.payload.cellIndex],
            amount: state.table[action.payload.id][action.payload.cellIndex].amount + 1
          })
      ])
      const newTable = [...state.table]
      return {
        ...state,
        table: newTable
      }
    }

    case SET_LIGHT_CELLS: {
      // const amount = action.payload.amount
      // const highlights = action.payload.highlights
      const rowId = action.payload.id
      const cellIndex = action.payload.cellIndex
      // const col = state.table.flat().length / state.table.length
      console.log(rowId, cellIndex)
      break

      // const lightArray = [...state.table].flat().map((item, index) => ({
      //   ...item,
      //   rowIndex: index,
      //   difference: Math.abs(item.amount - amount)
      // })).sort((a, b) => a.difference - b.difference).slice(0, highlights + 1)

      // state.table.map(cell => (
      //   lightArray.some(lightCell => lightCell.cellId === cell.cellId)
      //     ? { ...cell, isLight: true }
      //     : cell
      // ))

      // return {
      //   ...state,
      //   table: newTable
      // }
    }

    default:
      return state
  }
}

const initialState = {
  rows: '',
  columns: '',
  highlights: '',
  table: [],
  showTable: false,
  showButtons: false
}

const store = createStore(
  rootReducer,
  initialState,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store
