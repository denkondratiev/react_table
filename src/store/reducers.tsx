import { combineReducers } from 'redux'
import { ACTION_TYPES } from './actions'
import { Action, Params, RowsParams, CellsParams } from './actions'

const paramsReducer = (state: Params = {}, action: Action): Params => {
  const { SET_PARAMS } = ACTION_TYPES

  switch (action.type) {
    case SET_PARAMS:
      return action.payload
    default:
      return state
  }
}

function tableReducer (state: Array<string> = [], action: Action): Array<string> {
  const { SET_TABLE, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE:
      return action.payload
    case ADD_ROW:
      return [...state, action.payload.table]
    case REMOVE_ROW:
      return state.filter(id => id !== action.payload.lastRow)
    default:
      return state
  }
}

const rowsReducer = (state: RowsParams = {}, action: Action): RowsParams => {
  const { SET_ROWS, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_ROWS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.rows }
    case REMOVE_ROW:
      delete state[action.payload.lastRow]
      return state
    default:
      return state
  }
}


const cellsReducer = (state: CellsParams = {}, action: Action): CellsParams => {
  const { SET_CELLS, INCREMENT, ADD_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_CELLS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.cells }
    case INCREMENT:
      return {
        ...state,
        [action.payload]: {
          ...state[action.payload],
          amount: state[action.payload].amount + 1
        }
      }
    default:
      return state
  }
}

const buttonsReducer = (state: boolean = false, action: Action): boolean => {
  const { SHOW_BUTTONS } = ACTION_TYPES

  switch (action.type) {
    case SHOW_BUTTONS:
      return action.payload
    default:
      return state
  }
}

export const rootReducer = combineReducers({
  params: paramsReducer,
  table: tableReducer,
  rows: rowsReducer,
  cells: cellsReducer,
  buttons: buttonsReducer
})
