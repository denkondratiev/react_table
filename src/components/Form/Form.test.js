import React from 'react'
import { Provider } from 'react-redux'
import Form from '../Form/Form'
import { act, create } from 'react-test-renderer'
import { createStore } from 'redux'
import { rootReducer } from '../../store/reducers'

describe('testing form methods', () => {
  const store = createStore(rootReducer)
  let component

  act(() => {
    component = create(
      <Provider store={store}>
        <Form />
      </Provider>
    )
  })
  const root = component.root

  it('should correct render initial form', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('typed to form input', () => {
    const name = 'rowsAmount'
    const input = root.findByProps({ name })
    const event = { currentTarget: { name, value: 2 } }

    act(() => input.props.onChange(event))

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('form submit', () => {
    const form = root.findByProps({ id: 'form' })
    const withPrevent = { preventDefault: jest.fn() }

    act(() => form.props.onSubmit(withPrevent))

    expect(component.toJSON()).toMatchSnapshot()
  })
})
