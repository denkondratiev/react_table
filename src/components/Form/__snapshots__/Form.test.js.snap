// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing form methods form submit 1`] = `
<form
  className="form"
  id="form"
  onSubmit={[Function]}
>
  <input
    className="form-control false"
    name="rowsAmount"
    onChange={[Function]}
    placeholder="Rows..."
    type="number"
    value={2}
  />
  <div>
    Please add correct columns value
  </div>
  <input
    className="form-control error"
    name="columnsAmount"
    onChange={[Function]}
    placeholder="Columns..."
    type="number"
    value=""
  />
  <div>
    Please add correct highlight value
  </div>
  <input
    className="form-control error"
    name="lightsAmount"
    onChange={[Function]}
    placeholder="Highlight cells..."
    type="number"
    value=""
  />
  <button
    className="btn btn-dark"
    id="button-generate"
    type="submit"
  >
    Generate
  </button>
</form>
`;

exports[`testing form methods should correct render initial form 1`] = `
<form
  className="form"
  id="form"
  onSubmit={[Function]}
>
  <input
    className="form-control false"
    name="rowsAmount"
    onChange={[Function]}
    placeholder="Rows..."
    type="number"
    value=""
  />
  <input
    className="form-control false"
    name="columnsAmount"
    onChange={[Function]}
    placeholder="Columns..."
    type="number"
    value=""
  />
  <input
    className="form-control false"
    name="lightsAmount"
    onChange={[Function]}
    placeholder="Highlight cells..."
    type="number"
    value=""
  />
  <button
    className="btn btn-dark"
    id="button-generate"
    type="submit"
  >
    Generate
  </button>
</form>
`;

exports[`testing form methods typed to form input 1`] = `
<form
  className="form"
  id="form"
  onSubmit={[Function]}
>
  <input
    className="form-control false"
    name="rowsAmount"
    onChange={[Function]}
    placeholder="Rows..."
    type="number"
    value={2}
  />
  <input
    className="form-control false"
    name="columnsAmount"
    onChange={[Function]}
    placeholder="Columns..."
    type="number"
    value=""
  />
  <input
    className="form-control false"
    name="lightsAmount"
    onChange={[Function]}
    placeholder="Highlight cells..."
    type="number"
    value=""
  />
  <button
    className="btn btn-dark"
    id="button-generate"
    type="submit"
  >
    Generate
  </button>
</form>
`;
